stdlib/asm/crypto/stark/verifier.masm

```diff
@@ -0,0 +1,1439 @@
+#   7. Remainder commitment
+#   8. Intermidiate values.
+#   9. [lde_size, log2(lde_size), lde_domain_generator, 0] (same as 5 but for a different purpose)
+#   10.[gz1, gz0, z1, z0]
+#   11.[x, _, _, _] where x := offset * lde_domain_generator^query_index
+#   20. [0, 0, 0, 1] used for hmerge in case the number of elements to be hashed is not divisible by 8
+#   21. [0, 0, 0, 0] used for hmerge in case the number of elements to be hashed is divisible by 8
+#   22 -> 1999 Empty for now but should hold the public inputs in the future
+#   2000 -> 2015    auxiliary trace random coefficients
+#   2016 -> 2250    constraint composition coefficients
+#   2251    [(z^8)_1, (z^8)_0, z_1, z_0] where z = (z_0, z_1) is the OOD random element
+#   2252 -> 2292    current row of the OOD evaluation frame 
+#   2293 -> 2333    next row of the OOD evaluation frame 
+#   2334 -> 2337    value_i for i in 0..=7 where H(z) := \sum_0^7 value_i * z^i and H is the constraint compostion polynomial
+#   2338 -> 2463    Deep query coefficients
+#   2473 -> 2499    [0, query, evaluation_1, evaluation_0] for the 27 FRI queries
+#   2500 -> ptr_remainder - 1 where ptr_remainder := 2500 + 2 * number_of_layers
+#   ptr_remainder -> ptr_remainder + remainder_length / 2
+#   ptr_remainder + remainder_length / 2 + 1 -> ptr_remainder + remainder_length / 2 + length_of_remainder_polynomial/2
+#================================================
+
+# Notes: 
+# 1. The choice of the number of locals is based on an estimate and should be updated  
+# 2. Once `RandomCoin` has been redesigned, the way randomness is generated should be simplified.
+# 3. The periodic values for the OOD evaluation frame are not implemented. This should probably be
+#   delegated to the constraint evaluator.  
+# 4. Unless otherwise stated, `x`, `X`, `y` and `Y` denote "garbage" values on the stack. When computing
+#   the deep evaluation queries, `x` will denote `offset * g^index`.  
+export.verify.4222
+
+    # Create the initial seed for randomness generation
+    # The following expects the input stack to be of size 16
+    ## Proof context (except trace length)
+    push.1208027408
+    push.1
+    push.4294967295
+    push.132103
+    push.16             
+    push.8              
+    push.27             
+
+    ## trace length
+    adv_push.1
+
+    ## Compute log2(lde_domain_length) and save it for later use
+    dup
+    ### Multiply by blowup factor
+    mul.8
+    ### Compute log and store
+    exec.log2
+    loc_storew.1
+    dropw
+    
+    ## Hash proof context
+    push.1.0.0.0
+    #=> [0, 0, 0, 1, B, A, ProgRoot]
+    swapw.2
+    swapw
+    #=> [B, A, 0, 0, 0, 1, ProgRoot]
+    hperm
+    #=> [Y, Y, C, ProgRoot]
+
+    ## Hash i/o stack
+    swapw.3
+    swapw
+    #=> [Y, ProgRoot, C, Y, ]
+    adv_loadw
+    hperm
+    repeat.3
+    adv_loadw
+    swapw
+    adv_loadw
+    hperm
+    end
+    adv_loadw
+    swapw
+    dropw
+    push.1.0.0.0
+    hperm
+    swapdw
+    dropw
+    dropw
```

Comment:
A question to help me better understand your remark: what are the public inputs to the recursive MASM verifier?

Reply:
It is the public inputs defined [here](https://github.com/0xPolygonMiden/miden-vm/blob/main/air/src/lib.rs#L253). Basically, information about the program being executed (program hash + kernel) and information about initial and final state of the stack.

By the way, it does seem like we need to pass the hash of public inputs to the verifier - but mostly so that we can load the correct sequence of elements from the advice provider.

Reply:
Yes, that's how I am thinking about it. A hash of the public inputs is provided with the proof we are verifying in the VM. This hash is used to un-hash the public inputs that are provided through the advice stack.

Reply:
One more thing: having the trace length as well as the numbers of queries be inputs to the verification procedure seems a bit unclean to me. Should we provide all public inputs via the advice provider and only their hash via the input stack?

Reply:
I've been thinking about this: one argument for providing these inputs via the stack is that the calling procedure may want to have this info to compute proof security level.

An example here would be the batch kernel. This kernel would verify several transaction proofs, and we's want to make sure that these proofs are at some sufficient security level (i.e., don't have 3-bit security). The party which executes the batch kernel obviously knows security level of all transaction proofs. But the party which verifies correct execution of the batch kernel would not have this data. Thus, ideally, within the batch kernel itself we'd want to have the logic for computing proof security levels.

If we go with this, we'd actually need to provide 4 parameters to the verifier procedure:
- trace length.
- number of queries.
- blowup factor.
- grinding factor.

As all 4 of these are needed for conjectured security estimation.